### PR TITLE
Transaction Availability and Impact of Failures

### LINKED ISSUE #16072: [Docs] Availability of Transactions
### Description

- Add details of High Availability of Transaction in the Explore section
- Update `Distributed transactions` page with the Impact of Failures..

### PR DESCRIPTION
**Pull request by premkumr**
- Adds a new  High Availability of Transaction page in the Explore section under Continuous Availability
- Updates Distributed Transactions page with the Impact of Failures..

fixes #16072

@netlify /preview/explore/fault-tolerance/transaction-availability/
@netlify /preview/architecture/transactions/distributed-txns/

### CODE REVIEW

**Comment by premkumr** [comment]
@rkarthik007 , @prasad-rad , @polarweasel : Please review !!

**Comment by premkumr** [comment]
@kmuthukk : for review.

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 1]
```diff
@@ -85,3 +85,15 @@ After a transaction is committed, the following two fields are set:
...
+
+## Impact of Failures
+
+The Provisional Records are written to all the replicas of the tablets responsible for the keys being modified in a transaction. When a node with the tablet that has received the provisional records or is about to receive the provisional records fails, a new leader is elected for the tablet within a few seconds(`~2s`) as described in [Leader Failure](../../core-functions/high-availability/#tablet-peer-leader-failure). The transaction simply proceeds further with the newly elected leader. In this case, the time taken for the transaction to complete increases by the time taken for the leader election.
+
+The transaction manager (typically, the node the client is connected to) maintains the transaction-id to client mapping and sends heartbeats to the tablets containing the provisional records of the transaction. When the manager fails, the heartbeats stop and the provisional records remain uncommitted in the respective tablets.  Provisional records that do not get heartbeats become stale after certain time. Clients connected to the failed coordinator will receive an error message:
```
Comment (robertsami) [inline docs/content/preview/architecture/transactions/distributed-txns.md:93]: what is the transaction-id -> client mapping referring to? which clients?

Comment (robertsami) [reply docs/content/preview/architecture/transactions/distributed-txns.md:93]: why would the manager fail? what does that mean?

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 2]
```diff
@@ -85,3 +85,15 @@ After a transaction is committed, the following two fields are set:
...
+
+The transaction manager (typically, the node the client is connected to) maintains the transaction-id to client mapping and sends heartbeats to the tablets containing the provisional records of the transaction. When the manager fails, the heartbeats stop and the provisional records remain uncommitted in the respective tablets.  Provisional records that do not get heartbeats become stale after certain time. Clients connected to the failed coordinator will receive an error message: 
+
+```output.sql
+FATAL:  57P01: terminating connection due to unexpected postmaster exit
+FATAL:  XX000: Network error: recvmsg error: Connection refused
+```
+As the client is unaware of the transaction-id and the client to transaction-id mapping cannot be regenerated, it would be the responsibility of the client to retry the transaction. Other clients with transactions that were waiting on the transactions handled by the failed coordinator, will have to wait for the transaction's provisional records and locks to expire and then proceed normally.
```
Comment (robertsami) [inline docs/content/preview/architecture/transactions/distributed-txns.md:99]: we should note that when a coordinator does not receive a heartbeat from the client driving a transaction, it automatically cancels that transaction, so those provisional records woudl no longer block conflicting/competing transactions on the same keys

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+Insert some sample data and determine hashcode of the primary keys.
+
+```sql
+INSERT INTO txndemo SELECT id,10 FROM generate_series(1,5) AS id;
+```
+
+### Identifying the Key location
+For the purpose of the examples, you would need to identify which node holds a specific row, so that you
+can kill the correct node to see what is happening. For this one needs to get the hash code of the primary key using the `yb_hash_code(id)` function and co-relate with the `yb-admin` output to figure out where that 
+key is located. For the examples we will focus on the row with `k1=1`.
```
Comment (robertsami) [inline docs/content/preview/explore/fault-tolerance/transaction-availability.md:57]: why is this separated across lines?

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+```
+Notice that `127.0.0.2` is gone from the list and `127.0.0.3` is now the leader for the tablet that was in `127.0.0.2`(`7e2dfb66a..`)
+
+Commit the transaction.
+```output.sql
+yugabyte@yugabyte=# COMMIT;
+COMMIT
+Time: 6.243 ms
+```    
+We can see that even though the node `127.0.0.2` failed after receiving the provisional write, the transaction has succeeded.
```
Comment (robertsami) [inline docs/content/preview/explore/fault-tolerance/transaction-availability.md:124]: might be good to add a step where we read the table after commit and confirm the data is there

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+```output.sql
+yugabyte@yugabyte=# UPDATE txndemo set k2=30 where k1=1; COMMIT;
+UPDATE 1
+Time: 1728.246 ms (00:01.728)
+COMMIT
+Time: 2.964 ms
+```    
+You can see that even though the node `127.0.0.2` failed before receiving the provisional write, the transaction has succeeded. This is because a new leader (node:`127.0.0.3`) was quickly elected.
+    
+## Failure of Transaction Manager(Coordinator)
```
Comment (robertsami) [inline docs/content/preview/explore/fault-tolerance/transaction-availability.md:172]: do we have precedent for calling this "Transaction Manager"? i thought we call it "Transaction Coordinator" everywhere? let's stay consistent if so

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+```
+```output.sql
+ k1 | k2
+----+----
+  1 | 30
+(1 row)
+```
+
+The row did not get the intended value of `40` and still has the old value of `30`. The transaction did not go through. When the transaction coordinator fails before commit happens, the transaction is lost. At this point, it will be the application's responsibility to retry the transaction.
+
```
Comment (robertsami) [inline docs/content/preview/explore/fault-tolerance/transaction-availability.md:222]: is it worth showing an example where the client node fails and the rows it had provisional records on quickly become free to update in another session?

**Review by robertsami — CHANGES_REQUESTED** [review]

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 3 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:57]: Probably an erroneous <Enter> . But rendering is correct. Will fix it.

File: docs/content/preview/architecture/transactions/distributed-txns.md (hunk 1 above)
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md:93]: For the purpose of this page and other Availability pages, `fail` means, the node is unreachable either because of a crash or network failure.

The txn-id that points to the current txn in that connection. I was told that even though the txn-id is present in the status tablet and in the provisional records, the client will not be able to continue to the txn via connecting to another node as it would not know the txn-id.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 4 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:124]: Good point. I had felt that showing the `COMMIT` succeeding was enough. But yes, I'll add the confirmation of data.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 6 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:222]: It is certainly within the context, but techincally that would be client failure. This page is about server side failures. Let us ask the docs folks .. @aishwarya24 , @polarweasel , @prasad-rad - any suggestions here ?

File: docs/content/preview/architecture/transactions/distributed-txns.md (hunk 2 above)
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md]: done..

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 5 above)
Comment (polarweasel) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:181]: In the current 2.17 docs...

Transaction coordinator: 14 results in 8 files.
Transaction manager: 12 results in 7 files.

So really, we can call it whatever we want. One global search and replace will sort it out.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 6 above)
Comment (polarweasel) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:231]: Do you want to split this into two pages, one for node failures and one for client failures? That might be the best way to keep this page on-topic while adding some useful client-side examples.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 5 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:181]: This [diagram](https://docs.yugabyte.com/preview/architecture/transactions/transactional-io-path/#:~:text=The%20following%20diagram%20depicts%20the%20high%2Dlevel%20steps%20of%20a%20distributed%20write%2Donly%20transaction%2C%20not%20including%20any%20conflict%20resolution%3A)  says `Transaction Manager` . Not sure how many other images show the concept. Let discuss this soon and make a call.

Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md]: Lets discuss with @rahuldesirazu, @robertsami offline & decide on the naming convention . Lets keep that change separate from this diff.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 6 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:186]: I'll be addressing the client failures and retries in a separate page ..

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 7]
```diff
@@ -85,3 +85,17 @@ After a transaction is committed, the following two fields are set:
...
 - List of IDs of participating tablets: After a transaction commits, the final set of tablets to which the transaction has written is known. The tablet server managing the transaction sends this list to the transaction status tablet as part of the commit message, and the transaction status tablet makes sure that all participating tablets are notified of the transaction's committed status. This process might take multiple retries, and the transaction record can only be cleaned up after this is done.
+
+## Impact of Failures
+
+The Provisional Records are written to all the replicas of the tablets responsible for the keys being modified in a transaction. When a node with the tablet that has received the provisional records or is about to receive the provisional records fails, a new leader is elected for the tablet within a few seconds(`~2s`) as described in [Leader Failure](../../core-functions/high-availability/#tablet-peer-leader-failure). The transaction simply proceeds further with the newly elected leader. In this case, the time taken for the transaction to complete increases by the time taken for the leader election.
```
Comment (robertsami) [inline docs/content/preview/architecture/transactions/distributed-txns.md:91]:
```suggestion
Provisional records are written to all the replicas of the tablets responsible for the keys being modified in a transaction. When a node with the tablet that has received the provisional records or is about to receive the provisional records fails, a new leader is elected for the tablet within a few seconds(`~2s`) as described in [Leader Failure](../../core-functions/high-availability/#tablet-peer-leader-failure). The transaction simply proceeds further with the newly elected leader. In this case, the time taken for the transaction to complete increases by the time taken for the leader election.
```

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 8]
```diff
@@ -85,3 +85,17 @@ After a transaction is committed, the following two fields are set:
...
+
+When the transaction manager (typically, the node the client is connected to) does not receive a heartbeat from the client, it automatically cancels that transaction, so those provisional records would no longer block conflicting transactions waiting on the same keys. The transaction manager also maintains the transaction-id to client mapping and sends heartbeats to the transaction status tablet containing the information about the the transaction. When the manager fails, the heartbeats stop and the provisional records remain uncommitted in the respective tablets. Transactions that do not get heartbeats at the transaction status tablet become stale after certain time. Clients connected to the failed manager will receive an error message:
```
Comment (robertsami) [inline docs/content/preview/architecture/transactions/distributed-txns.md:93]: in this case i think its actually the coordinator who cancels the transaction if it doesn't receive a heartbeat from the pg client

> The transaction manager also maintains the transaction-id to client mapping and sends heartbeats to the transaction status tablet containing the information about the the transaction

i'm not sure what this sentence is trying to say, can you rephrase?

> When the manager fails, the heartbeats stop and the provisional records remain uncommitted in the respective tablets.

why would the manager fail? give an example

> Transactions that do not get heartbeats at the transaction status tablet become stale after certain time.

this is the same case as above, so maybe note that. also "stale after certain time" is imprecise -- we can explain how that happens in terms of these flags: https://github.com/yugabyte/yugabyte-db/blob/master/src/yb/tablet/transaction_coordinator.cc#L74

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 9]
```diff
@@ -85,3 +85,17 @@ After a transaction is committed, the following two fields are set:
...
+
+```output.sql
+FATAL:  57P01: terminating connection due to unexpected postmaster exit
+FATAL:  XX000: Network error: recvmsg error: Connection refused
+```
+
+As the client is unaware of the transaction-id and the client to transaction-id mapping cannot be regenerated, it would be the responsibility of the client to retry the transaction. Other clients with transactions that were waiting on the transactions handled by the failed manager, will have to wait for the transaction's provisional records and locks to expire and then proceed normally.
```
Comment (robertsami) [inline docs/content/preview/architecture/transactions/distributed-txns.md:100]:
```suggestion
As the client is unaware of the transaction-id, and since the client to transaction-id mapping cannot be regenerated, it would be the responsibility of the client to retry the transaction. Other clients with transactions that were blocked on provisional records written by the failed manager will have to wait for the transaction to be expired due to heartbeat timeout and can then proceed normally.
```

Comment (robertsami) [reply docs/content/preview/architecture/transactions/distributed-txns.md:100]: i posted this suggestion, but im not clear on a few things:
1. what is the client to transaction-id mapping you're referring to?
2. what do you mean when you say the client should retry the transaction? do you mean start a new transaction and try the same operations? if so, be more precise about that, because "retry the transaction" sounds like it may refer to trying to revive the already-failed transaction somehow

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+headcontent: Highly available and fault tolerant transactions
+menu:
+  preview:
+    identifier: transaction-availability-local
+    parent: fault-tolerance-1-macos
+    weight: 215
+type: docs
+---
+
+[Transactions](../../../architecture/transactions/distributed-txns/) are critical to many applications and need to work through different failure scenarios. YugabyteDB provides [high availability](../../../architecture/core-functions/high-availability/) (HA) of transactions by replicating the uncommitted values, a.k.a [provisional records](../../../architecture/transactions/distributed-txns/#provisional-records) across the [fault domains](../../../architecture/docdb-replication/replication/#fault-domains). The following examples demonstrate how YugabyteDB transactions survive common failure scenarios that could happen when a transaction is being processed. Some of the scenarios are:
```
Comment (robertsami) [inline docs/content/preview/explore/fault-tolerance/transaction-availability.md:15]: im still unclear on what we mean by "high availability of transactions". will defer reviewing the rest of this doc until we can sync on what this means

**Review by robertsami — CHANGES_REQUESTED** [review]

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    ```output
+     k1 | k2
+    ----+----
+      1 | 30
+    (1 row)
+    ```
+
+The transaction has succeeded: even though the node at `127.0.0.2` failed after receiving the provisional write, the value has been updated to `30`. The transaction succeeded because a new leader (the node at `127.0.0.3`) was quickly elected.
+
+## Transaction manager (coordinator) failure
```
[the remark this answers is not in the document]
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md]: done.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+    ```
+
+    ```output
+     k1 | k2
+    ----+----
+      1 | 30
+    (1 row)
+    ```
+
+The transaction has succeeded: even though the node at `127.0.0.2` failed after receiving the provisional write, the value has been updated to `30`. The transaction succeeded because a new leader (the node at `127.0.0.3`) was quickly elected.
```
[the remark this answers is not in the document]
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:241]: I'll change the wording. Will add details about the query layer waiting and resuming the transaction in the other page.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+
+    ```sh
+    yugabyted destroy --base_dir=/tmp/ybd1
+    yugabyted destroy --base_dir=/tmp/ybd2
+    yugabyted destroy --base_dir=/tmp/ybd3
+    ```
+
+## Conclusion
+
+Replication is a first-class feature in YugabyteDB and naturally handles node failures during transactions. During a transaction manager failure, even though provisional records are replicated, the maintains the correlation between the client and a transaction ID, of which the client is unaware.
```
[the remark this answers is not in the document]
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:340]: I've reworded the conclusion.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+type: docs
+---
+
+[Transactions](../../../architecture/transactions/distributed-txns/) are critical to many applications and need to work through different failure scenarios. YugabyteDB provides [high availability](../../../architecture/core-functions/high-availability/) (HA) of transactions by replicating the uncommitted values, a.k.a [provisional records](../../../architecture/transactions/distributed-txns/#provisional-records) across the [fault domains](../../../architecture/docdb-replication/replication/#fault-domains). The following examples demonstrate how YugabyteDB transactions survive common failure scenarios that could happen when a transaction is being processed. Some of the scenarios are:
+
+- The node that has received the provisional write fails (Handled by YugabyteDB)
+- The node that is about to receive the provisional write fails (Handled by YugabyteDB)
+- The transaction coordinator fails (Retry by client)
+
+<ul class="nav nav-tabs-alt nav-tabs-yb">
```
[the remark this answers is not in the document]
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md:21]: If I understand correctly, we try to pick the node responsible for the keys in the txn to also be the coordinator(for latency). That is why I avoided it, as we are actually explaining that case with failure of the node about to receive(or has received) the provisional write.

File: docs/content/preview/architecture/transactions/distributed-txns.md (hunk 9 above)
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md]: Made the appropriate changes.

File: docs/content/preview/explore/fault-tolerance/transaction-availability.md (hunk 10 above)
Comment (premkumr) [reply docs/content/preview/explore/fault-tolerance/transaction-availability.md]: Explained offline.

File: docs/content/preview/architecture/transactions/distributed-txns.md (hunk 8 above)
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md]: Explained offline and Made appropriate changes.

File: docs/content/preview/architecture/transactions/distributed-txns.md (hunk 7 above)
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md]: Accepted the changes

File: docs/content/preview/architecture/transactions/distributed-txns.md [hunk 15]
```diff
@@ -85,3 +85,15 @@ After a transaction is committed, the following two fields are set:
...
 - List of IDs of participating tablets: After a transaction commits, the final set of tablets to which the transaction has written is known. The tablet server managing the transaction sends this list to the transaction status tablet as part of the commit message, and the transaction status tablet makes sure that all participating tablets are notified of the transaction's committed status. This process might take multiple retries, and the transaction record can only be cleaned up after this is done.
+
+## Impact of Failures
+
+The Provisional Records are written to all the replicas of the tablets responsible for the keys being modified in a transaction. When a node with the tablet that has received the provisional records or is about to receive the provisional records fails, a new leader is elected for the tablet within a few seconds(`~2s`) as described in [Leader Failure](../../core-functions/high-availability/#tablet-peer-leader-failure). The transaction simply proceeds further with the newly elected leader. In this case, the time taken for the transaction to complete increases by the time taken for the leader election.
```
[the remark this answers is not in the document]
Comment (premkumr) [reply docs/content/preview/architecture/transactions/distributed-txns.md]: I've mentioned query layer will wait & retry . But intentionally leaving out  the term `rpc`

**Comment by rkarthik007** [comment]
Some comments:

* The TOC on the right feels weird, some items are unnecessarily nested inside others (for example: *Identify the key location* and *Clean Up* are subsections instead of top level sections). My suggestion is make *Clean Up* a top section, and create an initial section called *Prerequisites* - put *Setup* and *Identify location of a row* (NOTE: row, not key) as the two subsections inside this section. 
* We need to rename *Failure Scenario 1* and *Failure Scenario 2* to something the end user can understand. Same thing with *Transaction Manager Failure* as it implies a knowledge of the architecture
* The sections probably need diagrams to explain the scenarios
